src/brpc/rdma/rdma_endpoint.cpp

```diff
@@ -0,0 +1,1467 @@
+    if (_new_rq_wrs.fetch_add(num, butil::memory_order_relaxed) > _remote_window_capacity / 2) {
+        return SendImm(_new_rq_wrs.exchange(0, butil::memory_order_relaxed));
+    }
+    return 0;
+}
+
+int RdmaEndpoint::SendImm(uint32_t imm) {
+    if (imm == 0) {
+        return 0;
+    }
+
+    ibv_send_wr wr;
+    memset(&wr, 0, sizeof(wr));
+    wr.opcode = IBV_WR_SEND_WITH_IMM;
+    wr.imm_data = butil::HostToNet32(imm);
+    wr.send_flags |= IBV_SEND_SOLICITED;
+    wr.send_flags |= IBV_SEND_SIGNALED;
+
+    ibv_send_wr* bad = NULL;
+    if (ibv_post_send(_resource->qp, &wr, &bad) < 0) {
+        // We use other way to guarantee the Send Queue is not full.
+        // So we just consider this error as an unrecoverable error.
+        PLOG(WARNING) << "Fail to ibv_post_send";
+        return -1;
+    }
+    return 0;
+}
+
+ssize_t RdmaEndpoint::HandleCompletion(ibv_wc& wc) {
+    bool zerocopy = FLAGS_rdma_recv_zerocopy;
+    switch (wc.opcode) {
+    case IBV_WC_SEND: {  // send completion
+        // Do nothing
+        break;
+    }
+    case IBV_WC_RECV: {  // recv completion
+        // Please note that only the first wc.byte_len bytes is valid
+        if (wc.byte_len > 0) {
+            if (wc.byte_len < (uint32_t)FLAGS_rdma_zerocopy_min_size) {
+                zerocopy = false;
+            }
+            CHECK(_state != FALLBACK_TCP);
+            if (zerocopy) {
+                butil::IOBuf tmp;
+                _rbuf[_rq_received].cutn(&tmp, wc.byte_len);
+                _socket->_read_buf.append(tmp);
+            } else {
+                // Copy data when the receive data is really small
+                _socket->_read_buf.append(_rbuf_data[_rq_received], wc.byte_len);
+            }
+        }
+        if (wc.imm_data > 0) {
+            // Clear sbuf here because we ignore event wakeup for send completions
+            uint32_t acks = butil::NetToHost32(wc.imm_data);
+            uint32_t num = acks;
+            while (num > 0) {
+                _sbuf[_sq_sent++].clear();
+                if (_sq_sent == _sq_size - RESERVED_WR_NUM) {
+                    _sq_sent = 0;
+                }
+                --num;
+            }
+            butil::subtle::MemoryBarrier();
+
+            // Update window
+            uint32_t wnd_thresh = _local_window_capacity / 8;
+            if (_window_size.fetch_add(acks, butil::memory_order_relaxed) >= wnd_thresh
+                    || acks >= wnd_thresh) {
+                // Do not wake up writing thread right after _window_size > 0.
+                // Otherwise the writing thread may switch to background too quickly.
+                _socket->WakeAsEpollOut();
+            }
+        }
+        // We must re-post recv WR
+        if (PostRecv(1, zerocopy) < 0) {
+            return -1;
+        }
+        if (wc.byte_len > 0) {
+            SendAck(1);
+        }
+        return wc.byte_len;
+    }
+    default:
+        CHECK(false) << "This should not happen";
```

Comment:
```suggestion
        // Do not CHECK abort since wc.opcode could be IBV_WC_DRIVER2(136) which is
        // a bug already fixed by rdma-core.
        // FYI: https://github.com/linux-rdma/rdma-core/commit/4c905646de3e75bdccada4abe9f0d273d76eaf50
        LOG(WARNING) << "This should not happen, wc.opcode = " << wc.opcode;
```

Reply:
这个mlx驱动的bug会在什么情况下触发呢？如果确实是一个严重问题，仅仅抛出WARNING应该也是不行的。